api/agent/stats.go

```diff
@@ -2,58 +2,256 @@ package agent
 
 import (
 	"context"
-	"github.com/fnproject/fn/api/common"
+
+	"github.com/sirupsen/logrus"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
 )
 
-func StatsEnqueue(ctx context.Context) {
-	common.IncrementGauge(ctx, queuedMetricName)
-	common.IncrementCounter(ctx, callsMetricName)
+// TODO add some suga:
+// * hot containers active
+// * memory used / available
+
+func statsEnqueue(ctx context.Context) {
+	stats.Record(ctx, queuedMeasure.M(1))
+	stats.Record(ctx, callsMeasure.M(1))
 }
 
 // Call when a function has been queued but cannot be started because of an error
-func StatsDequeue(ctx context.Context) {
-	common.DecrementGauge(ctx, queuedMetricName)
+func statsDequeue(ctx context.Context) {
+	stats.Record(ctx, queuedMeasure.M(-1))
 }
 
-func StatsDequeueAndStart(ctx context.Context) {
-	common.DecrementGauge(ctx, queuedMetricName)
-	common.IncrementGauge(ctx, runningMetricName)
+func statsDequeueAndStart(ctx context.Context) {
+	stats.Record(ctx, queuedMeasure.M(-1))
+	stats.Record(ctx, runningMeasure.M(1))
 }
 
-func StatsComplete(ctx context.Context) {
-	common.DecrementGauge(ctx, runningMetricName)
-	common.IncrementCounter(ctx, completedMetricName)
+func statsComplete(ctx context.Context) {
+	stats.Record(ctx, runningMeasure.M(-1))
+	stats.Record(ctx, completedMeasure.M(1))
 }
 
-func StatsFailed(ctx context.Context) {
-	common.DecrementGauge(ctx, runningMetricName)
-	common.IncrementCounter(ctx, failedMetricName)
+func statsFailed(ctx context.Context) {
+	stats.Record(ctx, runningMeasure.M(-1))
+	stats.Record(ctx, failedMeasure.M(1))
 }
 
-func StatsDequeueAndFail(ctx context.Context) {
-	common.DecrementGauge(ctx, queuedMetricName)
-	common.IncrementCounter(ctx, failedMetricName)
+func statsDequeueAndFail(ctx context.Context) {
+	stats.Record(ctx, queuedMeasure.M(-1))
+	stats.Record(ctx, failedMeasure.M(1))
 }
 
-func StatsIncrementTimedout(ctx context.Context) {
-	common.IncrementCounter(ctx, timedoutMetricName)
+func statsTimedout(ctx context.Context) {
+	stats.Record(ctx, timedoutMeasure.M(1))
 }
 
-func StatsIncrementErrors(ctx context.Context) {
-	common.IncrementCounter(ctx, errorsMetricName)
+func statsErrors(ctx context.Context) {
+	stats.Record(ctx, errorsMeasure.M(1))
 }
 
-func StatsIncrementTooBusy(ctx context.Context) {
-	common.IncrementCounter(ctx, serverBusyMetricName)
+func statsTooBusy(ctx context.Context) {
+	stats.Record(ctx, serverBusyMeasure.M(1))
 }
 
 const (
+	// TODO we should probably prefix these with calls_ ?
 	queuedMetricName     = "queued"
-	callsMetricName      = "calls"
+	callsMetricName      = "calls" // TODO this is a dupe of sum {complete,failed} ?
```

Comment:
no, it's more like complete + failed + server_busy. The notion that 'failed' does not include server_busy I think it confusing, but preserved this for some time for UI consistency... (AKA UI won't show server too busy errors in failed stats. I think it should.. But anyways.)

Reply:
Unrelated with this PR, but I think the stats.go functions (especially after recent code reductions) do not provide much value anymore. They even make the code slightly harder to follow. I kind of want to break these "two-metric" functions here into "operate-on-one-metric" only functions or just get rid of them...

Reply:
that makes sense. `calls` stills seems superfluous then